sql/sort.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// orderBy constructs a sortNode based on the ORDER BY clause. Construction of
+// the sortNode might adjust the number of render targets in the scanNode if
+// any ordering expressions are specified.
+func (p *planner) orderBy(n *parser.Select, s *scanNode) (*sortNode, error) {
+	if n.OrderBy == nil {
+		return nil, nil
+	}
+
+	// We grab a copy of columns here because we might add new render targets
+	// below. This is the set of columns requested by the query.
+	columns := s.Columns()
+	var ordering []int
+
+	for _, o := range n.OrderBy {
+		index := 0
+
+		expr := o.Expr
+		if qname, ok := expr.(*parser.QualifiedName); ok {
+			if len(qname.Indirect) == 0 {
+				target := string(qname.Base)
+				for j, col := range columns {
+					if equalName(target, col) {
+						index = j + 1
+						break
+					}
+				}
+			}
+
+			if index == 0 {
+				if err := qname.NormalizeColumnName(); err != nil {
+					return nil, err
+				}
+				if qname.Table() == "" || equalName(s.desc.Alias, qname.Table()) {
+					for j, r := range s.render {
```

Comment:
I should add some more comments. The above loop over `s.columns` is checking to see if the qualified name matches one of the output columns. If we can't find a match, this loop is checking to see if the qualified name matches one of the existing render targets. This might also fail, in which case we add a new render target to the scanNode down on line 79.